Jump to python
==

자료형 값을 저장하는 공간, 변수
--

* 변수
  - 변수를 만들 때, =(assignment) 기호 사용
  - 파이썬은 변수에 저장된 값을 스스로 판단해 자료형의 타입을 지정함
  - 변수 이름 = 변수에 저장할 값

In [10]:
a = 1
b = "python"
c = [1, 2, 3]
d = (1, 2, 3)
e = {1, 2, 3}
f = {1:1, 2:2, 3:3}

In [17]:
type(a)

int

In [12]:
type(b)

str

In [13]:
type(c)

list

In [14]:
type(d)

tuple

In [15]:
type(e)

set

In [16]:
type(f)

dict

* 변수란?
  - 변수는 객체를 가리키는 것
  - 객체? 자료형의 데이터(값)

In [19]:
a = [1, 2, 3]
a

[1, 2, 3]

>  - [1, 2, 3]의 값을 가지는 리스트 데이터(객체)가 자동으로 메모리에 생성되고, 변수 a는 [1, 2, 3] 리스트가 저장된 메모리의 주소를 가리킴.

>  - 메모리? 컴퓨터가 프로그램에서 사용하는 데이터를 기억하는 공간

In [20]:
a = [1, 2, 3]
id(a)

140557266914688

>  - id 함수는 변수가 가리키고 있는 객체의 주소 값을 리턴하는 파이썬 내장 함수
>  - 변수 a가 가리키는 [1, 2, 3] 리스트의 주소 값은 140557266914688

* 리스트를 복사하고 싶을 때

In [21]:
a = [1, 2, 3]
b = a

# b와 a는 완전히 동일
# [1, 2, 3]이라는 리스트 객체를 참조하는 변수가 a 변수 1개에서 b 변수가 추가되어, 총 2개로 늘어남

In [22]:
id(a)

140557266591104

In [24]:
id(b)

# id(a) = id(b)

140557266591104

In [25]:
a is b

# a와 b가 가리키는 객체는 동일한가?
# 파이썬 명령어 is 는 동일한 객체를 가지고 있는지에 대해 판단

True

In [26]:
a[1] = 4
a

[1, 4, 3]

In [27]:
b

# a의 리스트 요소 변경 시, b도 함께 변경됨
# a와 b가 모두 동일한 리스트를 가리키고 있기 때문

[1, 4, 3]

* b 변수 생성 시, a 변수의 값을 가져오면서 a와 다른 주소를 가리키도록 만드는 방법

>  - [:] 이용

In [28]:
a = [1, 2, 3]
b = a[:]
a[1] = 4
a

[1, 4, 3]

In [29]:
b

# a 리스트 값을 바꾸어도 b 리스트에는 영향 끼치지 않음

[1, 2, 3]

>  - copy 모듈 이용

In [30]:
from copy import copy
a = [1, 2, 3]
b = copy(a)

In [31]:
b is a

# 두 변수의 값은 같지만, 서로 다른 객체를 가리키고 있음

False

In [32]:
a = [1, 2, 3]
b = a.copy()

* 변수를 만드는 여러 가지 방법

In [34]:
a, b = ('python', 'life')

# 튜플로 a, b 에 값 대입 가능

In [35]:
a

'python'

In [36]:
b

'life'

In [37]:
(a, b) = 'python', 'life'

In [38]:
[a,b] = ['python', 'life']

In [39]:
a = b = 'python'

In [40]:
a = 3
b = 5
a, b = b, a
a

5

In [41]:
b

3